### PR TITLE
fix(attention): align eager softcap mask dtype after upcast

### PR DESCRIPTION
**Pull request by glaziermag**
> [!IMPORTANT]
> Agent 4 A100 validation update (2026-05-13 UTC): classification `NOT_REPRODUCED` for the real runtime path and `SYNTHETIC` for the focused unit coverage; feasibility `FEASIBLE_NOW`. On A100 base `2d4ba4f16f61e5e18be085d0dd137bc95cba038a`, the real BF16 Gemma command `mistralrs-server --no-paged-attn plain -m unsloth/gemma-2-2b-it -d bf16 --max-seq-len 256 --max-batch-size 1` loaded, completed dummy run, and chat returned successfully; no dtype mismatch or NaN appeared in logs. PR head `2a3a5f67b086105e0642bdf82c2dec157eedd93f` also passed that runtime smoke and its two focused `softcap_attention` tests. Safe wording: this is a defensive eager-attention dtype alignment with BF16 score/mask tests; do not claim a reproduced NaN fix or actual before/after real-Gemma failure from this A100 run.

---

## Concrete issue fixed

When eager F16/BF16 attention scores are upcast to F32 for softcap and softmax, any attention mask/bias added afterward must be cast to the same dtype. Otherwise a real BF16 Gemma softcap path can fail with a dtype mismatch before generation starts.

## Fix

- Upcast eager F16/BF16 attention scores to F32 before softcap `tanh` and softmax.
- Keep those scores in F32 through the nonlinear operations.
- Cast the attention mask to the score dtype before `broadcast_add`.
- Cast probabilities back before multiplying by V.

## Antagonistic logic vetting result

The first version handled score upcasting but missed the mask/bias add. Cloud runtime testing found the concrete failure on a real Gemma 2 softcap model:

```text
prompt step - Model failed with error: dtype mismatch in add, lhs: F32, rhs: BF16
Error: model error: dtype mismatch in add, lhs: F32, rhs: BF16
```

This PR now includes a regression test for BF16 scores plus a BF16 mask.

## Tests

Commands run locally on branch head `2a3a5f67`:

```bash
cargo test -p mistralrs-core softcap_attention --lib
cargo fmt --all -- --check
git diff --check
cargo clippy --workspace --tests --examples -- -D warnings
```

Result: all passed locally.

Additional cloud compile/test check on 2026-04-29:

- VM: GCP `e2-standard-8` CPU VM `[redacted VM name]`, Rust `1.95.0`.
- Branch head tested: `2a3a5f67b`.

```bash
cargo test -q -p mistralrs-core softcap_attention --lib
```

Result: passed, 2 tests passed.

## Runtime validation

Cloud runtime validation on GCP L4:

- VM: `codex-l4-pr-vet-0428b`, `[redacted-region]`, `g2-standard-12`, 1x NVIDIA L4, CUDA toolkit 12.9 / driver 580.126.09.
- Model: `unsloth/gemma-2-2b-it` (ungated public Gemma 2 mirror).
- Relevant model config: `attn_logit_softcapping: Some(50.0)`, `final_logit_softcapping: Some(30.0)`, selected dtype BF16.

Before the mask-cast fix, branch head `523b9ad418bdeb07c93f867a0508e4c721a347d5` failed during the dummy run with the dtype mismatch above.

After the fix, branch head `2a3a5f67b086105e0642bdf82c2dec157eedd93f` passed:

```bash
cargo build --release --features cuda -p mistralrs --example text_models
RUST_LOG=info ./target-gemma/release/examples/text_models
```

Observed result: model loaded, dummy run completed, and chat generation completed without NaN or dtype errors.

## Relationship to the original NaN concern

This PR should be read as a tested real-Gemma softcap dtype fix. It is not a reproduced before/after fix for a separate softcap NaN report; the concrete issue reproduced here was the BF16 mask add exposed by score upcasting.

### CODE REVIEW

**Comment by glaziermag** [comment]
Agent 6 follow-up on existing A100 validation: closing this PR as not valid for the original A100-backed runtime claim. The recorded classification is `NOT_REPRODUCED` for the real runtime path and `SYNTHETIC` for the focused unit coverage, with feasibility `FEASIBLE_NOW`. The defensive dtype change may still be useful, but this PR should not remain open as a fix for the original runtime/NaN issue without a real failing base and passing head under the same conditions. A narrower synthetic/defensive PR can be opened if maintainers want that change.
